applications/nrf5340_audio/src/modules/sd_card_playback.c

```diff
@@ -0,0 +1,411 @@
+		return ret;
+	}
+	sd_card_playback_active = false;
+	return 0;
+}
+
+static void sd_card_playback_thread(void *arg1, void *arg2, void *arg3)
+{
+	int ret;
+	while (1) {
+		if (!sw_codec_is_initialized()) {
+			k_msleep(200);
+			continue;
+		}
+		k_sem_take(&m_sem_playback, K_FOREVER);
+		switch (playback_file_format) {
+		case SD_CARD_PLAYBACK_WAV:
+			ret = sd_card_playback_play_wav();
+			if (ret < 0) {
+				LOG_ERR("Wav playback err: %d", ret);
+			}
+			break;
+		case SD_CARD_PLAYBACK_LC3:
+			ret = sd_card_playback_play_lc3();
+			if (ret < 0) {
+				LOG_ERR("LC3 playback err: %d", ret);
+			}
+			break;
+		}
+	}
+}
+
+bool sd_card_playback_is_active(void)
+{
+	return sd_card_playback_active;
+}
+
+void sd_card_playback_wav(char *filename, uint32_t frame_duration_ms, uint8_t bit_depth,
+			  enum sd_playback_sample_rates sample_rate)
+{
+	playback_file_format = SD_CARD_PLAYBACK_WAV;
+	playback_file_name = filename;
+	wav_playback_cfg.frame_duration_ms = frame_duration_ms;
+	wav_playback_cfg.bit_depth = bit_depth;
+	wav_playback_cfg.sample_rate = sample_rate;
+	k_sem_give(&m_sem_playback);
+}
+
+int sd_card_playback_lc3(char *filename)
+{
+	if (!sw_codec_is_initialized()) {
+		LOG_ERR("Sw codec not initialized");
+		return -EACCES;
+	}
+	playback_file_format = SD_CARD_PLAYBACK_LC3;
+	playback_file_name = filename;
+	k_sem_give(&m_sem_playback);
+	return 0;
+}
+
+int sd_card_playback_mix_with_stream(void *const pcm_a, size_t pcm_a_size)
+{
+	int ret;
+	uint8_t pcm_b[pcm_frame_size];
+	size_t write_size = pcm_frame_size;
+
+	ret = sd_card_playback_buffer_set(pcm_b, &write_size);
+	if (ret < 0) {
+		LOG_ERR("Loading data into buffer err: %d", ret);
+		return ret;
+	}
+	if (write_size == pcm_frame_size) {
+		ret = pcm_mix(pcm_a, pcm_a_size, pcm_b, pcm_frame_size, B_MONO_INTO_A_STEREO_L);
+		if (ret < 0) {
+			LOG_ERR("Pcm mix err: %d", ret);
+		}
+	} else {
+		LOG_WRN("Size written does not equal requested size");
+	}
+	return 0;
+}
+
+int sd_card_playback_init(void)
+{
+	int ret;
+
+	sd_card_playback_thread_id = k_thread_create(
+		&sd_card_playback_thread_data, sd_card_playback_thread_stack,
+		CONFIG_SD_CARD_PLAYBACK_STACK_SIZE, (k_thread_entry_t)sd_card_playback_thread, NULL,
+		NULL, NULL, CONFIG_SD_CARD_PLAYBACK_THREAD_PRIORITY, 0, K_NO_WAIT);
+	ret = k_thread_name_set(sd_card_playback_thread_id, "sd_card_playback");
+	if (ret < 0) {
+		return ret;
+	}
+	return 0;
+}
+
+/* Shell functions */
+static int cmd_play_wav_file(const struct shell *shell, size_t argc, char **argv)
+{
+	if (!sw_codec_is_initialized()) {
+		shell_error(shell, "Sw codec not initialized");
+		return -EACCES;
+	}
+	char file_loc[MAX_PATH_LEN] = "";
+
+	strcat(file_loc, playback_file_path);
+	strcat(file_loc, argv[1]);
+	sd_card_playback_wav(file_loc, wav_playback_cfg.frame_duration_ms,
+			     wav_playback_cfg.bit_depth, SAMPLE_RATE_48K);
+	return 0;
+}
+
+static int cmd_play_lc3_file(const struct shell *shell, size_t argc, char **argv)
```

Comment:
Should the `sw_codec_is_initialized`-check be added here as well?

Reply:
Don't know if it is necessary as it is checked in the function it calls

Reply:
True. In that case you should make sure that this check is at the same level for both LC3 and WAV - right now the check is in `cmd_play` for WAV, but `sd_card_playback` for LC3. (I don't really have a preference for which of the two as long as it's the same for both)